### ISSUE TITLE
JsonSourceGenerator throwing exceptions after edit removing attribute

### ISSUE THREAD

**Issue by jasonmalinowski** [dotnet/roslyn#57455]
**Version Used**: 

**Steps to Reproduce**:

1. Launch VS under a debugger.
2. Set a breakpoint at https://github.com/dotnet/roslyn/blob/0734ecda9631c12798ef28ca3db87a2de43f6f65/src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs#L298

[remaining steps are in the debugee]

3. Create a new .NET 6 console app.
4. Expand the analyzers node, and observe System.Text.Json.SourceGeneration is not producing any files
5. Paste in the following contents:

```
class Person { }

internal partial class MyJsonContext : JsonSerializerContext
{
}
```
6. Observe still no generated files.
7. Add `[JsonSerializable(typeof(Person))]` before the MyJsonContext class.
8. Observe there are now generated files.
9. Delete the attribute
10. Observe that exceptions are now being thrown from the generator:

```
[3:37 PM] Jason Malinowski
SyntaxTree is not part of the compilation
Parameter name: syntaxTreeat Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetSemanticModel(SyntaxTree syntaxTree, Boolean ignoreAccessibility) in F:\Projects\Roslyn\src\Compilers\CSharp\Portable\Compilation\CSharpCompilation.cs:line 2256
at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CommonGetSemanticModel(SyntaxTree syntaxTree, Boolean ignoreAccessibility) in F:\Projects\Roslyn\src\Compilers\CSharp\Portable\Compilation\CSharpCompilation.cs:line 3547
at Microsoft.CodeAnalysis.Compilation.GetSemanticModel(SyntaxTree syntaxTree, Boolean ignoreAccessibility) in F:\Projects\Roslyn\src\Compilers\Core\Portable\Compilation\Compilation.cs:line 211
at System.Text.Json.SourceGeneration.JsonSourceGenerator.Parser.GetGenerationSpec(IEnumerable`1 classDeclarationSyntaxList)
at System.Text.Json.SourceGeneration.JsonSourceGenerator.Execute(Compilation compilation, ImmutableArray`1 contextClasses, SourceProductionContext sourceProductionContext)
at System.Text.Json.SourceGeneration.JsonSourceGenerator.<Initialize>b__4_3(SourceProductionContext spc, ValueTuple`2 source)
at Microsoft.CodeAnalysis.UserFunctionExtensions.<>c__DisplayClass3_0`2.<WrapUserAction>b__0(TInput1 input1, TInput2 input2) in F:\Projects\Roslyn\src\Compilers\Core\Portable\SourceGeneration\UserFunction.cs:line 66
--- End of stack trace from previous location ---
at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
```

My hunch here is when we are removing the attributes, we're still keeping around the old one for later generation passes which is now stale.

**Comment by jasonmalinowski** [comment]
@jcouv this was marked for 17.3 but missed the cutoff. Was this something you did want in for 17.3?

**Comment by jcouv** [comment]
Thanks for calling this out @jasonmalinowski. From discussion with Jared, I'll cherry-pick the fix to 17.3 (PR https://github.com/dotnet/roslyn/pull/62249)
